src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,278 @@
+        }
+
+        /// <summary>
+        /// Subscribes to the underlying <see cref="IObservable{T}"/> with the
+        /// transformation specified by this instance.
+        /// <br/><br/>
+        /// Disconnection requests via the returned <see cref="IDisposable"/> interface
+        /// are passed to the underlying <see cref="IObservable{T}"/> and also used
+        /// to signal pending asynchronous tasks that cancellation has been requested
+        /// and also used to prevent further event notifications.
+        /// </summary>
+        public IDisposable Subscribe(IObserver<TOut> observer)
+        {
+            IDisposable? subscription = null;
+            var newObserver = new Observer(observer, _transformNext, _transformError, () => subscription?.Dispose());
+            subscription = _source.Subscribe(newObserver);
+            return newObserver;
+        }
+
+        private class Observer : IObserver<TIn>, IDisposable
+        {
+            private CancellationTokenSource? _cancellationTokenSource = new();
+            private readonly CancellationToken _token;
+            //create a queue so that events will be sent in order
+            private readonly Queue<QueueEvent> _queue = new();
+            private readonly IObserver<TOut> _observer;
+            private readonly Func<TIn, CancellationToken, ValueTask<TOut>> _transformNext;
+            private readonly Func<Exception, CancellationToken, ValueTask<Exception>> _transformError;
+            private Action? _disposeAction;
+
+            public Observer(IObserver<TOut> observer, Func<TIn, CancellationToken, ValueTask<TOut>> transformNext, Func<Exception, CancellationToken, ValueTask<Exception>> transformError, Action disposeAction)
+            {
+                _token = _cancellationTokenSource.Token;
+                _observer = observer;
+                _disposeAction = disposeAction;
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<Exception>
+                _transformError = async (exception, token) => await transformError(exception, token).ConfigureAwait(false);
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<TOut>
+                _transformNext = async (data, token) => await transformNext(data, token).ConfigureAwait(false);
+            }
+
+            public void OnNext(TIn value)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                Queue(QueueType.Data, _transformNext(value, _token), default);
+            }
+
+            public void OnError(Exception error)
+            {
+                if (_token.IsCancellationRequested)
+                    return;
+                Queue(QueueType.Error, default, _transformError(error, _token));
+            }
+
+            public void OnCompleted() => Queue(QueueType.Completion, default, default);
```

Comment:
Why without `_token.IsCancellationRequested` check?

Reply:
Because the first line in `Queue` performs this check so it is redundant.  The other methods it makes a difference because it would prevent the transformations from executing.

Reply:
Understood.